src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs

```diff
@@ -751,114 +741,219 @@ private void ReturnReadBufferIfEmpty()
+            // DecryptData will decrypt in-place and modify these to point to the actual decrypted data, which may be smaller.
+            _decryptedBytesOffset = _internalOffset;
+            _decryptedBytesCount = frameSize;
+            SecurityStatusPal status;
+            lock (_handshakeLock)
+            {
+                ThrowIfExceptionalOrNotAuthenticated();
+                status = _context!.Decrypt(new ReadOnlySpan<byte>(_internalBuffer, _internalOffset, frameSize), _internalBuffer.AsSpan(_decryptedBytesOffset), out decryptedOffset, out decryptedCount);
+                _decryptedBytesCount = decryptedCount;
+                if (decryptedCount > 0)
+                {
+                    _decryptedBytesOffset = _internalOffset + decryptedOffset;
+                }
+
+                if (status.ErrorCode == SecurityStatusPalErrorCode.Renegotiate)
+                {
+                    // The status indicates that peer wants to renegotiate. (Windows only)
+                    // In practice, there can be some other reasons too - like TLS1.3 session creation
+                    // of alert handling. We need to pass the data to lsass and it is not safe to do parallel
+                    // write any more as that can change TLS state and the EncryptData() can fail in strange ways.
+
+                    // To handle this we call DecryptData() under lock and we create TCS waiter.
+                    // EncryptData() checks that under same lock and if it exist it will not call low-level crypto.
+                    // Instead it will wait synchronously or asynchronously and it will try again after the wait.
+                    // The result will be set when ReplyOnReAuthenticationAsync() is finished e.g. lsass business is over.
+                    // If that happen before EncryptData() runs, _handshakeWaiter will be set to null
+                    // and EncryptData() will work normally e.g. no waiting, just exclusion with DecryptData()
 
-                    SecurityStatusPal status;
-                    lock (_handshakeLock)
+
+                    if (_sslAuthenticationOptions!.AllowRenegotiation || SslProtocol == SslProtocols.Tls13)
                     {
-                        status = DecryptData();
-                        if (status.ErrorCode == SecurityStatusPalErrorCode.Renegotiate)
-                        {
-                            // The status indicates that peer wants to renegotiate. (Windows only)
-                            // In practice, there can be some other reasons too - like TLS1.3 session creation
-                            // of alert handling. We need to pass the data to lsass and it is not safe to do parallel
-                            // write any more as that can change TLS state and the EncryptData() can fail in strange ways.
+                        // create TCS only if we plan to proceed. If not, we will throw in block bellow outside of the lock.
```

Comment:
typo: "bellow" => "below"

Reply:
Also, it says we will throw outside the lock, but I don't see a throw here... am I missing it?

Reply:
comment updated. The general logic did not change but placement did.